numba/cuda/tests/cudapy/test_libdevice.py

```diff
@@ -0,0 +1,186 @@
+
+        np.testing.assert_array_equal(frac_expect, fracres)
+        np.testing.assert_array_equal(exp_expect, expres)
+
+    def test_sad(self):
+        # Test return of a scalar from a libdevice function
+        x = np.arange(0, 200, 2)
+        y = np.arange(50, 150)
+        z = np.arange(15, 115)
+        r = np.zeros_like(x)
+
+        cufunc = cuda.jit(use_sad)
+        cufunc[4, 32](r, x, y, z)
+
+        np.testing.assert_array_equal(np.abs(x - y) + z, r)
+
+
+# A template for generating tests of compiling calls to libdevice functions.
+# The purpose of the call and assignment of the return variables is to ensure
+# the actual function implementations are not thrown away resulting in a PTX
+# implementation that only contains the ret instruction - this may hide certain
+# errors.
+function_template = """\
+from numba.cuda import libdevice
+
+def pyfunc(%(pyargs)s):
+    ret = libdevice.%(func)s(%(funcargs)s)
+    %(retvars)s = ret
+"""
+
+
+def make_test_call(libname):
+    """
+    Generates a test function for each libdevice function.
+    """
+
+    def _test_call_functions(self):
+        # Strip off '__nv_' from libdevice name to get Python name
+        apiname = libname[5:]
+        apifunc = getattr(libdevice, libname[5:])
+        retty, args = functions[libname]
+        sig = create_signature(retty, args)
+
+        # Construct arguments to the libdevice function. These are all
+        # non-pointer arguments to the underlying bitcode function.
+        funcargs = ", ".join(['a%d' % i for i, arg in enumerate(args) if not
+                              arg.is_ptr])
+
+        # Arguments to the Python function (`pyfunc` in the template above) are
+        # the arguments to the libdevice function, plus as many extra arguments
+        # as there are in the return type of the libdevice function - one for
+        # scalar-valued returns, or the length of the tuple for tuple-valued
+        # returns.
+        if isinstance(sig.return_type, (types.Tuple, types.UniTuple)):
+            # Start with the parameters for the return values
+            pyargs = ", ".join(['r%d' % i for i in
+                                range(len(sig.return_type))])
+            # Add the parameters for the argument values
+            pyargs += ", " + funcargs
+            # Generate the unpacking of the return value from the libdevice
+            # function into the Python function return values (`r0`, `r1`,
+            # etc.).
+            retvars = ", ".join(['r%d[0]' % i for i in
+                                 range(len(sig.return_type))])
+        else:
+            # Scalar return is a more straightforward case
+            pyargs = "r0, " + funcargs
+            retvars = "r0[0]"
+
+        # Create the string containing the function to compile
+        d = { 'func': apiname,
+              'pyargs': pyargs,
+              'funcargs': funcargs,
+              'retvars': retvars }
+        code = function_template % d
```

Comment:
NOTE: I spot checked a few of these. Code gen looks fine.

Reply:
Many thanks!